mesonbuild/mdist.py

```diff
@@ -80,7 +80,36 @@ def is_git(src_root: str) -> bool:
     Checks if meson.build file at the root source directory is tracked by git.
     It could be a subproject part of the parent project git repository.
     '''
-    return quiet_git(['ls-files', '--error-unmatch', 'meson.build'], src_root)[0]
+    if quiet_git(['ls-files', '--error-unmatch', 'meson.build'], src_root)[0]:
+        return True
+
+    if os.path.exists(os.path.join(src_root, '.git')):
+        msg = 'Source tree looks like it may be a git repo, '
+        if not GIT:
+            msg += 'but git is not installed!'
+            if 'GITLAB_CI' in os.environ:
+                msg += ' This is a gitlab bug.'
+        else:
+            msg += 'but git returned a failure. '
+            p, oe = quiet_git(['status'], src_root)
+            if 'dubious ownership' in oe:
+                # For a few years now, git has absolved itself of the responsibility to implement
+                # robust, safe software. Instead of detecting the signs of a problematic scenario,
+                # they have chosen to consider many legitimate and reasonable use cases as "dangerous",
+                # and implemented the number one threat to security worldwide: alert fatigue. Having
+                # done so, they then washed their hands of the matter and permanently tabled the
+                # notion of adding fine-grained detection. This is not just useless, it is *worse*
+                # than useless.
+                #
+                # In our case, the error is triply meaningless since we are already executing build
+                # system commands from the same directory. Either way, reject the notion that git is
+                # well designed or that its error messaging is a valid approach to the problem space.
+                msg += 'This is a bug in git itself, please set `git config --global safe.directory "*"`'
```

Comment:
As much as I find the whole `safe.directory` fiasco to be a, well, fiasco... We should probably be diplomatic in our public code comments, and the message itself pretty much sums the whole thing up. Maybe just:
```txt
This kind of warning from git mix legitimately problematic configurations with those that are not in one kind of error message.
```

Reply:
The flip side is we may end up with people litigating our advice on the grounds that it is "bad for security to recommend disabling security features".

I figured that most people aren't going to look in our code comments to begin with, and I wanted to make it clear to future people trying to figure out why this warning exists, that the reason for my advice is because I genuinely do not consider it to be a security feature in the first place.

Reply:
Note: we agreed on IRC that this was fine to go ahead.